File: logic/validation/validate.js
const { errorHandler } = require("../classes/ErrorHandler");


const validateAargs = () => {
    // TODO check length
    if (process.argv.length !== 2) {
        errorHandler.showUsage();
    }
    return true;
};

const validateData = (line, place) => {
    const errors = [];
    for (let i = 0; i < line.length; i++) {
        const char = line.charAt(i);
        if (char !== char.toUpperCase()) {
            errors.push(`Wrong character ${char} in ${place}`)
        }
    }
    return errors;
};

const validateChars = (split, line) => {
    const errors = [];
    for (let item of split) {
         // TODO check regexp
        let wrongChars = item.split(/[A-Z+()^!|?]+/).join('');
        for (let i = 0; i < wrongChars.length; i++) {
            const char = wrongChars.charAt(i);
            errors.push(`Wrong character ${char} in ${line}`)
        }
    }
    return errors;
};

const validateLine = (line) => {
    const errors = [];
    if (line[0] === '=' || line[0] === '?') {
        if (line[0] === '=') {
            errors.push(...validateData(line.slice(1), 'facts'))
        }
        if (line[0] === '?') {
            errors.push(...validateData(line.slice(1), 'queries'))
        }
    } else {
        if (line.includes('<=>')) {
            const split = line.split('<=>');
            if (split.length !== 2) {
                errors.push(`Rule ${line} is not valid`);
            } else {
                errors.push(...validateChars(split, line));
            }
        } else if (line.includes('=>')) {
            const split = line.split('=>');
            if (split.length !== 2) {
                errors.push(`Rule ${line} is not valid`);
            } else {
                errors.push(...validateChars(split, line));
            }
        } else {
            errors.push(`Rule ${line} is not valid`);
        }
    }
    if (errors.length) {
        return errors;
    } else {
        return false;
    }
};

const validateKnownData = (facts, queries, factFlag) => {
    const errors = [];
    if (!facts.length && !factFlag) {
        errors.push('No facts');
    }
    if (!queries.length) {
        errors.push('No queries');
    }
    if (errors.length) {
        return errors;
    }
    return false;
};

File: logic/validation/validate.py
import sys, re
from classes.ErrorHandler import error_handler


def validate_args():
	if len(sys.argv) != 2:
		error_handler.show_usage()
	return True		# todo


def check_characters(line):
	pass


def validate_data(line, place):
	errors = []
	for char in line:
		# print line
		if not char.isupper():
			errors.append('Wrong character "{}" in {}'.format(char, place))
	return errors


def validate_chars(split, line):
	errors = []
	for item in split:
		wrong_chars = ''.join(re.split('[A-Z+()^!|?]+', item))
		for char in wrong_chars:
			errors.append('Wrong character "{}" in "{}"'.format(char, line))
	return errors


def validate_line(line):
	errors = []

	if line[0] == '=' or line[0] == '?':
		if line[0] == '=':
			errors += validate_data(line[1:], 'facts')

		if line[0] == '?':
			errors += validate_data(line[1:], 'queries')

	else:
		if '<=>' in line:
			split = line.split('<=>')
			if len(split) != 2:
				errors.append('Rule "{}" is not valid'.format(line))
			else:
				errors += validate_chars(split, line)

		elif '=>' in line:
			split = line.split('=>')
			if len(split) != 2:
				errors.append('Rule "{}" is not valid'.format(line))
			else:
				errors += validate_chars(split, line)

		else:
			errors.append('Rule "{}" is not valid'.format(line))

	if len(errors):
		return errors
	return False


def validate_known_data(facts, queries, fact_flag):
	errors = []
	if len(facts) == 0 and not fact_flag:
		errors.append('No facts')

	if len(queries) == 0:
		errors.append('No queries')

	if len(errors):
		return errors
	return False
